app/javascript/views/cryptoWalletServices/CryptoWalletServices.jsx

```diff
@@ -0,0 +1,250 @@
+import * as React from "react";
+import { injectIntl } from "react-intl";
+import ErrorBoundary from "../../components/errorBoundary/ErrorBoundary";
+import CryptoWalletOption from './CryptoWalletOption'
+import Select from 'react-select'
+import axios from "axios";
+import routes from "../routes";
+import bs58 from 'bs58'
+import Web3 from 'web3';
+
+
+class CryptoWalletServices extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.channel = props.channel;
+    
+    this.state = {
+      solOptions: [],
+      ethOptions: [],
+      currentSolAddress: null,
+      currentEthAddress: null,
+      isLoading: true,
+      errorText: null,
+    };
+  }
+
+  componentDidMount() {
+    this.loadData();
+  }
+
+  // Helper functions
+  formatOptionData(response, chain) {
+    const options = response.data.filter( address => address.chain === chain)
+                      .map( address => { return { value: address, label: address.address}})
+    options.push({label: 'Add new wallet', value: {newAddress: chain}});
+    return options;
+  }
+
+  findCurrentAddress(chain, channelResponse, allResponse) {
+    const current = channelResponse.data.filter( address => address.chain === chain)[0]
+    if (current) {
+      const found = allResponse.data.filter( address => address.chain === chain )
+        .find( address => address.id === current.crypto_address_id)
+      return found ? { label: this.formatCryptoAddress(found.address), value: found } : null
+    }
+    return null;
+  }
+
+  formatCryptoAddress(address) {
+    return `${address.slice(0,6)}****${address.slice(-4)}`;
+  }
+
+  // setup the dropdowns
+  loadData = () => {
+    this.setState({ isLoading: true });
+    axios.get(routes.publishers.cryptoAddresses.index).then((response) => {
+      const newState = {
+        isLoading: false,
+        errorText: null,
+      };
+      newState.solOptions = this.formatOptionData(response, 'SOL')
+      newState.ethOptions = this.formatOptionData(response, 'ETH')
+      
+      axios.get(routes.publishers.cryptoAddressForChannels.index.replace('{channel_id}', this.channel.id)).then((channelResponse) => {
+        newState.currentSolAddress = this.findCurrentAddress('SOL', channelResponse, response)
+        newState.currentEthAddress = this.findCurrentAddress('ETH', channelResponse, response)
+
+        this.setState({ ...newState });
+      });
+    });
+  };
+
+  // crypto connection functions
+  connectSolanaAddress = async () => {
+    if (!window.solana) {
+      return false;
+    }
+
+    const results = await window.solana.connect();
+    if (results.publicKey) {
+      const pub_key = results.publicKey
+
+      const possibleMatch = this.state.solOptions.filter(sol => sol.value.address === pub_key);
+      if (possibleMatch.length > 0) {
+        this.updateAddress(possibleMatch[0]);
+        return;
+      }
+
+      const nonce = Date.now().toString();
```

Comment:
Utilizing the date as the message here would result in an active MITM attack (because they'd be able to see the message being sent to the creators service) being able to claim ownership of an address that doesn't belong to them. In this case, we'd want to be able to generate the message for them server side rather than on the front end client and then compare the signed value to the originally generated message.

If you want a good example of how this is being done with brave talk in the front end, you can take a look at it here: https://github.com/brave/brave-talk/blob/main/src/components/web3/api.ts#L178-L222